gnc/sub8_perception/include/sub8_pcl/torpedo_board.hpp

```diff
@@ -20,41 +22,106 @@
   double image_proc_scale;
+  sub::Contour left_corners, right_corners;
+  sub::ImageWithCameraInfo left_most_recent, right_most_recent;
+
 private:
-  ros::NodeHandle nh;
-  ros::ServiceServer service;
+  // Callbacks
+  bool detection_activation_switch(
+      sub8_perception::TBDetectionSwitch::Request &req,
+      sub8_perception::TBDetectionSwitch::Response &resp);
+  void left_image_callback(const sensor_msgs::ImageConstPtr &image_msg_ptr,
+                           const sensor_msgs::CameraInfoConstPtr &info_msg_ptr);
+  void
+  right_image_callback(const sensor_msgs::ImageConstPtr &image_msg_ptr,
+                       const sensor_msgs::CameraInfoConstPtr &info_msg_ptr);
+
+  // Detection / Processing
+  void run();
+  void determine_torpedo_board_position();
+  void segment_board(const cv::Mat &src, cv::Mat &dest, cv::Mat &dbg_img,
+                     bool draw_dbg_img = false);
+  bool find_board_corners(const cv::Mat &segmented_board, sub::Contour &corners,
+                          bool draw_dbg_left = true);
 
+  // ROS
+  ros::NodeHandle nh;
+  ros::ServiceServer detection_switch;
+  ros::ServiceClient pose_client;
   image_transport::CameraSubscriber left_image_sub, right_image_sub;
   image_transport::ImageTransport image_transport;
   image_transport::Publisher debug_image_pub;
   image_geometry::PinholeCameraModel left_cam_model, right_cam_model;
 
-  sub::Contour left_corners, right_corners;
-  sub::ImageWithCameraInfo left_most_recent;
-  sub::ImageWithCameraInfo right_most_recent;
+  // Torpedo Board detection will be attempted when true
+  bool active;
 
+// Frames will be considered synchronized if their stamp difference is less than
+// this (in seconds)
+#if __cplusplus > 199711L
+  static constexpr double sync_thresh = 0.25;
+#else
   static const double sync_thresh = 0.25;
+#endif
 
+  // Goes into sequential id for pos_est srv request
+  long long int run_id = 0;
+
+  // RVIZ
   sub::RvizVisualizer rviz;
 
+  // DBG images will be generated and published when true
+  const bool generate_dbg_img;
   cv::Mat debug_image;
   cv::Rect upper_left, upper_right, lower_left, lower_right;
-  const bool generate_dbg_img;
+};
+
+class TorpedoBoardReprojectionCost {
+public:
+  TorpedoBoardReprojectionCost(cv::Matx34d &proj_L, cv::Matx34d &proj_R,
+                               std::vector<cv::Point> &corners_L,
+                               std::vector<cv::Point> &corners_R);
+  ~TorpedoBoardReprojectionCost();
+
+  template <typename T>
+  bool operator()(const T *const x, const T *const y, const T *const z,
+                  const T *const yaw, T *residual) const;
+
+private:
+  static std::vector<cv::Point> getProjectedCorners(double center_x,
+                                                    double center_y,
+                                                    double center_z, double yaw,
+                                                    cv::Matx34d &proj_matrix);
+
+#if __cplusplus > 199711L
```

Comment:
Compiling pcl from source with c++11 flags will fix it

- Jake

> On Apr 28, 2016, at 10:17 PM, David Santiago Soto <notifications@github.com> wrote:
> 
> In gnc/sub8_perception/include/sub8_pcl/torpedo_board.hpp:
> 
> > +  TorpedoBoardReprojectionCost(cv::Matx34d &proj_L, cv::Matx34d &proj_R,
> > +                               std::vector<cv::Point> &corners_L,
> > +                               std::vector<cv::Point> &corners_R);
> > +  ~TorpedoBoardReprojectionCost();
> > +
> > +  template <typename T>
> > +  bool operator()(const T *const x, const T *const y, const T *const z,
> > +                  const T *const yaw, T *residual) const;
> > +
> > +private:
> > +  static std::vector<cv::Point> getProjectedCorners(double center_x,
> > +                                                    double center_y,
> > +                                                    double center_z, double yaw,
> > +                                                    cv::Matx34d &proj_matrix);
> > +
> > +#if __cplusplus > 199711L
> @jpanikulam Never mind, it compiles fine, but at runtime, theres a pcl bug that causes a segfault before the node is even initialized. Apparently its been patched in the latest release though.
> 
> Going back to compiling with c++03. (10 minute compile times again!!!)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>